### PR TITLE
ParallelFlowExecutor service by Yurii Kotsiuba

### PR DESCRIPTION
**Pull request by jurok3x**
My implementation of ParallelFlowExecutor.

### CODE REVIEW

File: src/main/java/executor/service/service/ParallelFlowExecutorService.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package executor.service.service;
+
+public interface ParallelFlowExecutorService {
+    void run();
```
Comment (OlexTul) [inline src/main/java/executor/service/service/ParallelFlowExecutorService.java:4]: Why did you change the name of the method, this contradicts the contract.

File: src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        this.proxyQueue = new LinkedBlockingQueue<>(CAPACITY);
+    }
+
+    @Override
+    public void run() {
+        threadPoolExecutor.execute(() -> scenarioListener.execute(getScenarioHandler()));
+        threadPoolExecutor.execute(() -> proxyClient.execute(getProxyHandler()));
+        threadPoolExecutor.execute(this::runExecution);
+    }
+
```
Comment (OlexTul) [inline src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java:44]: In my opinion, here, we should still stick to the Java documentation and implement the **shutdown()** method as a good practice.

File: src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                proxyQueue.put(proxy);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        };
+    }
+
+    private void runExecution() {
+        try {
+            while (true) {
```
Comment (OlexTul) [inline src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java:67]: Here you should provide a correct exit from the loop **while{}**.

File: src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                Scenario scenario = scenarioQueue.take();
+                ProxyConfigHolder proxy = proxyQueue.take();
+                threadPoolExecutor.execute(() -> executionService.execute(scenario, proxy));
+            }
+        } catch (InterruptedException e) {
+            System.err.printf("Error while running Thread: %s", e.getMessage());
```
Comment (OlexTul) [inline src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java:73]: It would be better to use logging here.

**Review by OlexTul — CHANGES_REQUESTED** [review]

File: src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java (hunk 4 above)
Comment (Sokol3783) [reply src/main/java/executor/service/service/impl/ParallelFlowExecutorServiceImpl.java:73]: I'm agree

**Review by hromovyi-kostiantyn — APPROVED** [review]

**Review by Sokol3783 — APPROVED** [review]

**Review by Mercurialf — APPROVED** [review]

**Pull request closed without merging**